lib/capistrano/tasks/withrsync.rake

```diff
@@ -83,16 +83,18 @@ namespace :rsync do
   desc 'Sync to deployment hosts from local'
   task sync: :'rsync:stage' do
     last_rsync_to = nil
-    roles(:all).each do |role|
-      run_locally do
-        user = "#{role.user}@" if !role.user.nil?
-        rsync_options = "#{fetch(:rsync_options).join(' ')}"
-        rsync_from = "#{fetch(:rsync_src)}/"
-        rsync_to = "#{user}#{role.hostname}:#{fetch(:rsync_dest_fullpath) || release_path}"
-
-        unless rsync_to == last_rsync_to
-          execute :rsync, rsync_options, rsync_from, rsync_to
-          last_rsync_to = rsync_to
+    release_roles(:all).each do |role|
+      unless Capistrano::Configuration.env.filter(role).empty?
```

Comment:
`ENV['HOSTS' || 'ROLES']` are recognized in https://github.com/capistrano/capistrano/blob/master/lib/capistrano/configuration.rb#L94-L101

Codes like `Capistrano::Configuration.env.filter(roles(:all))` is used in capistrano build-in recipes to filter servers along environment variables.

And, `role` can be both a `Server` instance or Array of `Server`s.

(I wanted to add some test, but it is toooo difficult...)